### PR TITLE
Make interpolate_2d handle datetime64 correctly

### PR DESCRIPTION
**Pull request by jbrockmendel**
Broken off from #27626 because I decided that is poorly scoped.

### CODE REVIEW

**Review by gfyoung — CHANGES_REQUESTED** [review]
Fix looks relatively straightforward, just need tests and `whatsnew`

**Comment by jbrockmendel** [comment]
Test added, unnecessary piece removed.  No release note needed since no user-facing changes.

File: pandas/core/missing.py [hunk 1]
```diff
@@ -490,6 +491,10 @@ def interpolate_2d(
     if ndim == 1:
         values = values[0]
 
+    if orig_values.dtype.kind == "M":
+        # convert float back to datetime64
+        values = values.astype(orig_values.dtype)
```
Comment (TomAugspurger) [inline pandas/core/missing.py]: Can this fail if not all the values are filled (e.g. with `limit`)?

**Review by TomAugspurger — COMMENTED** [review]
LGTM overall. I noticed the following while testing locally

```python
In [1]: import pandas as pd

In [2]: dti = pd.date_range("2015-04-05", periods=5, tz="US/Central")
   ...: ser = pd.Series(dti)
   ...: ser.iloc[1:4] = pd.NaT
   ...: ser.astype('datetime64[ns]').to_frame().interpolate(method="pad", limit=2)
Out[2]:
                    0
0 2015-04-05 05:00:00
1                 NaT
2                 NaT
3                 NaT
4 2015-04-09 05:00:00
```

That happens on both master and this PR. Do you know if we have an issue for it? The important bits are

1. `datetime64[ns]`-dtype. datetime with TZ seems fine
2. DataFrame.interpolate. Series[datetime64[ns]].interpolate seems fine

File: pandas/core/missing.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/missing.py]: I don't see why it would.  This is functionally the same as what we're doing now, just in a more-correct place.

**Comment by jbrockmendel** [comment]
> datetime64[ns]-dtype. datetime with TZ seems fine
> DataFrame.interpolate. Series[datetime64[ns]].interpolate seems fine

I'm not sure I understand your example.  Is it relevant that you passed tz to date_range?  You subsequently astype-d it away.

Best guess: DatetimeTZBlock.fillna dispatches to DatetimeArray.fillna, so doesn't go directly through here.

**Comment by jbrockmendel** [comment]
@jreback gentle ping.  Between this and the quantile PR, we're about to be about to be able to rip out a bunch of _try_coerce_result code

File: pandas/core/missing.py (hunk 1 above)
Comment (jreback) [reply pandas/core/missing.py]: this likely should be for .kind in ['M', 'm']

Comment (jreback) [reply pandas/core/missing.py]: or even better this should be is_extension_dtype then cast back

Comment (jbrockmendel) [reply pandas/core/missing.py]: its not tz case we're catching here, it's regular datetime64

**Review by TomAugspurger — APPROVED** [review]
Thoughts @jreback?

File: pandas/core/missing.py (hunk 1 above)
Comment (jreback) [reply pandas/core/missing.py]: k

**Comment by jreback** [comment]
thanks

**Pull request merged**